pkg/ruler/rulestore/bucket_client.go

```diff
@@ -0,0 +1,312 @@
+		err = userBucket.Iter(ctx, prefix, func(group string) error {
+			group = strings.TrimSuffix(strings.TrimPrefix(group, prefix), delim)
+			decodedGroup, err := base64.URLEncoding.DecodeString(group)
+			if err != nil {
+				return err
+			}
+			groupList = append(groupList, &rules.RuleGroupDesc{
+				User:      userID,
+				Namespace: namespace,
+				Name:      string(decodedGroup),
+			})
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return groupList, nil
+}
+
+// LoadRuleGroups loads rules for each rule group in the map.
+// Parameter with groups to load *MUST* be coming from one of the List methods.
+// Reason is that some implementations don't do anything, since their List method already loads the rules.
+func (b *BucketRuleStore) LoadRuleGroups(ctx context.Context, groupsToLoad map[string]rules.RuleGroupList) error {
+	ch := make(chan *rules.RuleGroupDesc)
+
+	// Given we store one file per rule group. With this, we create a pool of workers that will
+	// download all rule groups in parallel. We limit the number of workers to avoid a
+	// particular user having too many rule groups rate limiting us with the object storage.
+	g, gCtx := errgroup.WithContext(ctx)
+	for i := 0; i < b.loadConcurrency; i++ {
+		g.Go(func() error {
+			for gr := range ch {
+				if gr == nil {
+					continue
+				}
+
+				user, namespace, group := gr.GetUser(), gr.GetNamespace(), gr.GetName()
+				if user == "" || namespace == "" || group == "" {
+					return fmt.Errorf("invalid rule group: user=%q, namespace=%q, group=%q", user, namespace, group)
+				}
+
+				gr, err := b.getRuleGroup(gCtx, user, namespace, group, gr) // reuse group pointer from the map.
+				if err != nil {
+					level.Error(b.logger).Log("msg", "failed to get rule group", "namespace", namespace, "name", group, "user", user)
+					return err
+				}
+
+				if user != gr.User || namespace != gr.Namespace || group != gr.Name {
+					return fmt.Errorf("mismatch between requested rule group and loaded rule group, requested: user=%q, namespace=%q, group=%q, loaded: user=%q, namespace=%q, group=%q", user, namespace, group, gr.User, gr.Namespace, gr.Name)
+				}
+			}
+
+			return nil
+		})
+	}
+
+outer:
+	for _, gs := range groupsToLoad {
+		for _, g := range gs {
+			select {
+			case <-gCtx.Done():
+				break outer
+			case ch <- g:
+				// ok
+			}
+		}
+	}
+	close(ch)
+
+	return g.Wait()
+}
+
+func (b *BucketRuleStore) GetRuleGroup(ctx context.Context, userID string, namespace string, group string) (*rules.RuleGroupDesc, error) {
+	return b.getRuleGroup(ctx, userID, namespace, group, nil)
+}
+
+func (b *BucketRuleStore) SetRuleGroup(ctx context.Context, userID string, namespace string, group *rules.RuleGroupDesc) error {
+	userBucket := newRuleStoreUserBucketClient(userID, b.bucket)
+	data, err := proto.Marshal(group)
+	if err != nil {
+		return err
+	}
+	return userBucket.Upload(ctx, generateRuleObjectKey(namespace, group.Name), bytes.NewReader(data))
```

Comment:
Please use `bytes.NewBuffer(data)` instead of `bytes.NewReader` as bucket `Upload` function tries to find the size of the data before upload, and it cannot do it for `bytes.Reader`. (See `objstore.TryToGetSize` for details)

Reply:
👍